[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70216
* #70373
* #76999
* __->__ #75995
* #70479
* #76776

Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

One behavioral change is that all ChildDataPipes will reset as soon as a new generation of iterator is created (previously reset only happens when a new iterator is created from a child DataPipe has been read, and no reset when it hasn't been read).

The implementation takes place within `_typing.py` and `combining.py` See the new unit test within `test_datapipe.py` for an example of the desired behavior. And https://github.com/pytorch/data/issues/45 for discussion.

## BC-breaking Note:
* The change here is closely related to #70479, please read that PR's BC-breaking note first.
* Each `ChildDataPipe` (output of `fork`, `demux`, `unzip`) can only have one active iterator at a time, as described in the previous PR's BC-breaking note. When a new generation of iterators is created, the buffer of the main DataPipe will reset and all ChildDataPipes from the previous generation will become invalid. The code example below illustrates the behavior:

**1.11.0 or before**:
```python
source_dp = IterableWrapper(range(10))
cdp1, cdp2 = source_dp.fork(num_instances=2)
it1, it2 = iter(cdp1), iter(cdp2)
list(it1)  # [0, ..., 9]
list(it2)  # [0, ..., 9]
it1, it2 = iter(cdp1), iter(cdp2)
it3 = iter(cdp1)  # Basically share the same reference as `it1`, doesn't reset because `cdp1` hasn't been read since reset
next(it1)  # returns 0, no error
next(it2)  # returns 0, no error
next(it3)  # returns 1
# The next line resets all ChildDataPipe because `cdp2` has started reading
it4 = iter(cdp2)
next(it3)  # returns 0
list(it4)  # [0, ..., 9]
```

**This PR**:
```python
source_dp = IterableWrapper(range(10))
cdp1, cdp2 = source_dp.fork(num_instances=2)
it1, it2 = iter(cdp1), iter(cdp2)
list(it1)  # [0, ..., 9]
list(it2)  # [0, ..., 9]
it1, it2 = iter(cdp1), iter(cdp2)
it3 = iter(cdp1)  # This invalidates `it1` and `it2`
next(it1)  # raises an error
next(it2)  # raises an error
next(it3)  # returns 0
# The next line should not invalidate anything, as there was no new iterator created
# for `cdp2` after `it2` was invalidated
it4 = iter(cdp2)
next(it3)  # returns 1, no error here
list(it4)  # [0, ..., 9]
```
* Again, if you need to reference the same `ChildDataPipe` multiple times, please apply `.fork()` on the `ChildDataPipe` as described in the [BC-breaking note of the previous PR](#70479).

Differential Revision: [D36156460](https://our.internmc.facebook.com/intern/diff/D36156460)

NivekT committed:
[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"




[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

See the unit test for example. And https://github.com/pytorch/data/issues/45 for discussion.


[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

See the unit test for example. And https://github.com/pytorch/data/issues/45 for discussion.


[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

See the unit test for example. And https://github.com/pytorch/data/issues/45 for discussion.


[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156460).

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

See the unit test for example. And https://github.com/pytorch/data/issues/45 for discussion.

Differential Revision: [D36156460](https://our.internmc.facebook.com/intern/diff/D36156460)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

One behavioral change is that all ChildDataPipes will reset as soon as a new generation of iterator is created (previously reset only happens when a new iterator is created from a child DataPipe has been read, and no reset when it hasn't been read).

See the unit test for example. And https://github.com/pytorch/data/issues/45 for discussion.

Differential Revision: [D36156460](https://our.internmc.facebook.com/intern/diff/D36156460)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156460).

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

One behavioral change is that all ChildDataPipes will reset as soon as a new generation of iterator is created (previously reset only happens when a new iterator is created from a child DataPipe has been read, and no reset when it hasn't been read).

See the unit test for example. And https://github.com/pytorch/data/issues/45 for discussion.

Differential Revision: [D36156460](https://our.internmc.facebook.com/intern/diff/D36156460)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156460).

NivekT commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -196,25 +193,42 @@ class _ChildDataPipe(IterDataPipe):
-            self.main_datapipe.reset()
-        # We want to separate the code for reset and yield, so that 'reset' exeutes before __next__ is called
+        if not self.main_datapipe.is_every_instance_exhausted():
+            warnings.warn("Some child DataPipes are not exhausted when __iter__ is called. We are resetting "
+                          "the buffer and each child DataPipe will read from the start again.", UserWarning)
```
This is currently causing a lot of warnings in our tests. It may causes too many warnings for users too if they happen to have the need to reset before reading everything. I think we have the following options:
1. Remove the warning entirely
2. Raises once, then not again (have a global variable that switches)
3. Option to silence this through a global API?
4. Keep this as it is

NivekT replied:
I can confirm through the tests within `torchvision` that it raises a lot of warning, so we definitely should not keep it as it is.

ejguan replied:
I support to silence all warnings within TorchData via a global API. For users, this warning is good to indicate their wrong implementation or some potential bugs because this won't happen a lot. But, for `torchvision` tests, it's reasonable to disable it as the test was introduced intentionally.

pmeier replied:
IIUC, this warning should only pop up if we actually are doing something dubious in the first place, right? @NivekT what datasets are giving you this warning? Should we simply fix our datasets instead of fixing the warning behavior?

pmeier replied:
> Raises once, then not again (have a global variable that switches)

This is the default Python behavior: 

```py
import warnings

def warn():
    warnings.warn("You should only see this once!")

for _ in range(5):
    warn()
```

The reason you are seeing this warning over and over in our test suite is that `pytest` does something similar to `warnings.simplefilter("always")`.

NivekT replied:
I re-test `torchvision` with my latest changes (after I moved the warning to the correct location). I'm no longer seeing any warnings.

NivekT commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -223,6 +237,33 @@ def __len__(self):
         yield from self.main_datapipe.get_next_element_by_instance(self.instance_id)
 
+    # This method is called by `hook_iterator` in `_typing.py`.
+    def _set_main_datapipe_valid_iterator_id(self) -> int:
```
Let me know if the explanation here is unclear. Also see the docstring/example within `class ChildDataPip`.

NivekT commented:
Note - I have tested the changes of this stack of PRs up to this one against `torchvision` (I ran `pytest test/test_prototype_builtin_datasets.py`, and all the tests pass as long as I disable the warnings here:

https://github.com/pytorch/pytorch/pull/75995/files#r868508119

cc: @pmeier

pmeier commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -338,9 +379,6 @@ def get_next_element_by_instance(self, instance_id: int):
-        return self.instance_started[instance_id]
-
     def is_every_instance_exhausted(self) -> bool:
         return self.main_datapipe_exhausted and all(not child_buffer for child_buffer in self.child_buffers)
```
Is this still correct? Why are we checking if the input datapipe is exhausted? This leads to being warned in very simple cases where this shouldn't be happening:

```py
from torchdata.datapipes.iter import IterableWrapper, Demultiplexer, Zipper

dp = IterableWrapper(range(6))
dps = Demultiplexer(dp, 3, lambda x: x % 3)
dp = Zipper(*dps)

next(iter(dp))
```

NivekT replied:
Thanks for pointing that out. I suspected something is off. 

The method is still correct but needs to be moved to a different place (right before `reset()` is called).

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

One behavioral change is that all ChildDataPipes will reset as soon as a new generation of iterator is created (previously reset only happens when a new iterator is created from a child DataPipe has been read, and no reset when it hasn't been read).

The implementation takes place within `_typing.py` and `combining.py` See the new unit test within `test_datapipe.py` for an example of the desired behavior. And https://github.com/pytorch/data/issues/45 for discussion.

Differential Revision: [D36156460](https://our.internmc.facebook.com/intern/diff/D36156460)

[ghstack-poisoned]

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -195,25 +192,42 @@ class _ChildDataPipe(IterDataPipe):
     will pass its instance_id to get the next value from its main DataPipe.
 
+    Note:
+        ChildDataPipe, like all other IterDataPipe, follows the singler iterator per IterDataPipe constraint.
```
snit: single

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -829,30 +831,30 @@ def test_fork_iterdatapipe(self):
-        self.assertEqual(list(range(10)), output2)
+        with self.assertRaisesRegex(RuntimeError, r"iterator has been invalidated"):
+            for i, n2 in enumerate(dp2):
+                output2.append(n2)
```
snit: replace next 3 lines with `pass`

NivekT replied:
The invalidation happens in the middle of the loop during the `if` condition. Therefore, it cannot be changed to `pass` and therefore will still be elements in `output2` that can be checked against.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -981,18 +983,18 @@ def test_demux_iterdatapipe(self):
-                i1 = iter(dp1)
+        with self.assertRaisesRegex(RuntimeError, r"iterator has been invalidated"):
+            for i, n2 in enumerate(dp2):
+                output2.append(n2)
```
snit: `pass`

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -2346,6 +2348,30 @@ def __iter__(self):
+        it4 = iter(cdp2)
+        self.assertEqual(1, next(it3))  # An error shouldn't be raised here
+        self.assertEqual(list(range(10)), list(it4))
+
```
Would be nice to see test for something like:
```python
         source_dp: IterDataPipe = dp.iter.IterableWrapper(range(10))
        cdp1, cdp2 = source_dp.fork(num_instances=2)
        it1, it2 = iter(cdp1), iter(cdp2)
        self.assertEqual(list(range(10)), list(it1))
        self.assertEqual(list(range(10)), list(it2))
        it1, it2 = iter(cdp1), iter(cdp2)
        it3 = iter(source_dp)  # This should impact it1 and it2
        with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
            next(it1)
```

NivekT replied:
Added

VitalyFedyunin reviewed (commented):
Overall good, but lets try to dive deeper (see if it breaks).

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

One behavioral change is that all ChildDataPipes will reset as soon as a new generation of iterator is created (previously reset only happens when a new iterator is created from a child DataPipe has been read, and no reset when it hasn't been read).

The implementation takes place within `_typing.py` and `combining.py` See the new unit test within `test_datapipe.py` for an example of the desired behavior. And https://github.com/pytorch/data/issues/45 for discussion.

Differential Revision: [D36156460](https://our.internmc.facebook.com/intern/diff/D36156460)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

One behavioral change is that all ChildDataPipes will reset as soon as a new generation of iterator is created (previously reset only happens when a new iterator is created from a child DataPipe has been read, and no reset when it hasn't been read).

The implementation takes place within `_typing.py` and `combining.py` See the new unit test within `test_datapipe.py` for an example of the desired behavior. And https://github.com/pytorch/data/issues/45 for discussion.

Differential Revision: [D36156460](https://our.internmc.facebook.com/intern/diff/D36156460)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

One behavioral change is that all ChildDataPipes will reset as soon as a new generation of iterator is created (previously reset only happens when a new iterator is created from a child DataPipe has been read, and no reset when it hasn't been read).

The implementation takes place within `_typing.py` and `combining.py` See the new unit test within `test_datapipe.py` for an example of the desired behavior. And https://github.com/pytorch/data/issues/45 for discussion.

Differential Revision: [D36156460](https://our.internmc.facebook.com/intern/diff/D36156460)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Enforcing single valid iterator for IterDataPipes multiple DataPipes as outputs"


Similar to the last PR in the stack, this PR only allow there to be one active iterator for each `ChildDataPipe`. Creating a new iterator will invalidate all previous, related iterators of `ChildDataPipe`.

One behavioral change is that all ChildDataPipes will reset as soon as a new generation of iterator is created (previously reset only happens when a new iterator is created from a child DataPipe has been read, and no reset when it hasn't been read).

The implementation takes place within `_typing.py` and `combining.py` See the new unit test within `test_datapipe.py` for an example of the desired behavior. And https://github.com/pytorch/data/issues/45 for discussion.

Differential Revision: [D36156460](https://our.internmc.facebook.com/intern/diff/D36156460)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156460).
